p2p/src/network/rpc/p2p_network_rpc_reducer.rs

```diff
@@ -1,33 +1,83 @@
+use mina_p2p_messages::{
+    rpc,
+    rpc_kernel::{
+        MessageHeader, PayloadBinprotReader as _, QueryHeader, ResponseHeader, RpcMethod,
+        RpcQueryReadError, RpcResponseReadError,
+    },
+    v2,
+    versioned::Ver,
+};
+use openmina_core::{error, fuzz_maybe, fuzzed_maybe, Substate};
+use redux::Dispatcher;
 
-use crate::{Limit, P2pLimits};
+use crate::{
+    channels::rpc::{
+        BestTipWithProof, P2pChannelsRpcAction, P2pRpcRequest, P2pRpcResponse,
+        StagedLedgerAuxAndPendingCoinbases,
+    },
+    connection::outgoing::P2pConnectionOutgoingInitOpts,
+    disconnection::{P2pDisconnectionAction, P2pDisconnectionReason},
+    Data, Limit, P2pLimits, P2pNetworkState, P2pNetworkYamuxAction, PeerId,
+};
 
 use self::p2p_network_rpc_state::P2pNetworkRpcError;
 
 use super::*;
 
 impl P2pNetworkRpcState {
-    pub fn reducer(
-        &mut self,
+    pub fn reducer<State, Action>(
+        mut state_context: Substate<Action, State, P2pNetworkState>,
         action: redux::ActionWithMeta<&P2pNetworkRpcAction>,
         limits: &P2pLimits,
-    ) {
-        match action.action() {
-            P2pNetworkRpcAction::Init { incoming, .. } => {
-                self.is_incoming = *incoming;
+    ) -> Result<(), String>
+    where
+        State: crate::P2pStateTrait,
+        Action: crate::P2pActionTrait<State>,
+    {
+        let rpc_state = state_context
+            .get_substate_mut()?
+            .find_rpc_state_mut(action.action())
+            .ok_or_else(|| format!("RPC state not found for action: {:?}", action.action()))?;
+
+        let (action, meta) = action.split();
+        match action {
+            P2pNetworkRpcAction::Init {
+                incoming,
+                addr,
+                peer_id,
+                stream_id,
+            } => {
+                rpc_state.is_incoming = *incoming;
+
+                let dispatcher = state_context.into_dispatcher();
+                dispatcher.push(P2pNetworkRpcAction::OutgoingData {
+                    addr: *addr,
+                    peer_id: *peer_id,
+                    stream_id: *stream_id,
+                    data: Data::from(RpcMessage::Handshake.into_bytes()),
+                    fin: false,
+                });
+                Ok(())
             }
-            P2pNetworkRpcAction::IncomingData { data, .. } => {
-                self.buffer.extend_from_slice(data);
+            P2pNetworkRpcAction::IncomingData {
+                data,
+                addr,
+                peer_id,
+                stream_id,
+            } => {
+                rpc_state.buffer.extend_from_slice(data);
                 let mut offset = 0;
                 // TODO(akoptelov): there shouldn't be the case where we have multiple incoming messages at once (or at least other than heartbeat)
                 loop {
-                    let buf = &self.buffer[offset..];
+                    let buf = &rpc_state.buffer[offset..];
```

Comment:
Lets take the opportunity to change indexing for `get(offset..)` and `bug_condition` for the `None` case